videosink/handler.go

```diff
@@ -44,8 +57,63 @@ func (d *Display) configFromQuery(values url.Values) (imageConfig, error) {
 }
 
 type client struct {
+	disp *Display
+
 	refresh   chan struct{}
 	terminate chan struct{}
+
+	mostRecent time.Time
+}
+
+func newClient(d *Display) *client {
+	return &client{
+		disp:      d,
+		refresh:   make(chan struct{}, 1),
+		terminate: make(chan struct{}, 1),
+	}
+}
+
+// Wait until the next frame should be sent, either because the keep-alive
+// interval has passed or a change has been made to the buffer and the rate
+// limit isn't violated.
+func (c *client) waitNext(ctx context.Context) bool {
+	earliestFrameAt := c.mostRecent.Add(c.disp.minFrameInterval)
+	latestFrameAt := c.mostRecent.Add(c.disp.keepAliveInterval)
+
+	var rateLimitTimer *time.Timer
+	var keepAliveTimer = time.NewTimer(time.Until(latestFrameAt))
+
+	defer stopAndDrain(keepAliveTimer)
+
+	for {
+		var rateLimit <-chan time.Time
+
+		if rateLimitTimer != nil {
+			rateLimit = rateLimitTimer.C
+		}
+
+		select {
+		case <-rateLimit:
+			return true
+
+		case <-c.refresh:
+			if remaining := time.Until(earliestFrameAt); remaining <= 0 {
+				return true
+			} else if rateLimitTimer == nil {
+				rateLimitTimer = time.NewTimer(remaining)
+				defer stopAndDrain(rateLimitTimer)
```

Comment:
You could make rateLimit a function local variable (instead of being inside the for loop) if you would set its value here. I think it'd be more readable.

Reply:
Good idea; done.